libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,747 @@
+    if( packetId == MQTT_PACKET_ID_INVALID )
+    {
+        *pIdStatus = ID_INVALID;
+        index = recordLen;
+    }
+    else
+    {
+        *pIdStatus = ID_UNKNOWN;
+        for( index = 0; index < recordLen; index++ )
+        {
+            if( records[ index ].packetId == packetId )
+            {
+                *pIdStatus = ID_KNOWN;
+                *pQos = records[ index ].qos;
+                *pCurrentState = records[ index ].publishState;
+                break;
+            }
+        }
+    }
+    return index;
+}
+
+static MQTTStatus_t _deleteRecord( MQTTPubAckInfo_t * records,
+                                   size_t recordLen,
+                                   size_t index )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    if( index < recordLen )
+    {
+        records[ index ].packetId = MQTT_PACKET_ID_INVALID;
+        records[ index ].qos = MQTTQoS0;
+        records[ index ].publishState = MQTTInvalidState;
+    }
+    else
+    {
+        /* Out of bounds error. */
+        status = MQTTBadParameter;
+    }
+    return status;
+}
+
+static MQTTStatus_t _addRecord( MQTTPubAckInfo_t * records,
+                                size_t recordLen,
+                                uint16_t packetId,
+                                MQTTQoS_t qos,
+                                MQTTPublishState_t publishState )
+{
+    MQTTStatus_t status = MQTTNoMemory;
+    int index = 0;
+    size_t availableIndex = recordLen;
+    if( packetId == MQTT_PACKET_ID_INVALID || qos == MQTTQoS0 )
+    {
+        status = MQTTBadParameter;
+    }
+    else
+    {
+        /* Start from end so first available index will be populated. */
+        for( index = recordLen - 1; index >= 0; index-- )
+        {
+            if( records[ index ].packetId == MQTT_PACKET_ID_INVALID )
+            {
+                availableIndex = index;
+            }
+            else if( records[ index ].packetId == packetId )
+            {
+                /* Collision. */
+                status = MQTTStateCollision;
+                availableIndex = recordLen;
+                break;
+            }
+            else
+            {
+                /* Empty else clause. */
+            }   
+        }
+    }
+
+    if( availableIndex < recordLen )
+    {
+        records[ availableIndex ].packetId = packetId;
+        records[ availableIndex ].qos = qos;
+        records[ availableIndex ].publishState = publishState;
+        status = MQTTSuccess;
+    }
+
+    return status;
+}
+
+static MQTTStatus_t _updateRecord( MQTTPubAckInfo_t * records,
+                                   size_t recordLen,
+                                   size_t recordIndex,
+                                   uint16_t packetId,
+                                   MQTTQoS_t qos,
+                                   MQTTPublishState_t newState,
+                                   bool shouldStore,
+                                   bool shouldDelete )
```

Comment:
Having separate parameters for `shouldStore` and `shouldDelete` means that someone can construct a call to this function which doesn't make sense, e.g. `shouldStore == 1 && shouldDelete == 1`.  Why not create an `enum` for action type, and pass that here?

```c
typedef enum
{
   NO_ACTION = 0,
   SHOULD_STORE = 1,
   SHOULD_DELETE =2
}   MQTTRecordAction_t
```
This would also let you rewrite this logic with a more readable `switch` pattern.

Reply:
Thank you for the suggestion, and I will update it. Originally this was part of _MQTT_UpdateState(), but was moved out to reduce complexity, hence the two booleans that were meant to be mutually exclusive. I agree an enum would be better here.

Reply:
Refactored to prevent this function from doing too many things.